### DISCUSSION TITLE
Zephyr Bluetooth DFU Emulation (NRF52840)

### DISCUSSION THREAD

**Discussion by vChavezB** [zephyrproject-rtos/zephyr#81521] [Show and tell]
I have been testing out the [`bumble`](https://github.com/google/bumble) HCI Controller Stack from Google with Renode and Zephyr.  I made a proof of concept to emulate completley DFU over a virtual HCI Controller (bleak +bumble) with renode.

My use case for this application is to develop high level functional tests for custom DFUs with MCUBoot.

Wanted to show a short demo of the application. On the left side is the zephyr logs from the emulated firmware on renode. On the right side I am using [pyrenode](https://github.com/antmicro/pyrenode3) + [`smpclient`](https://github.com/intercreate/smpclient) to communicate with the Zephyr SMP Server.  The demo connects to the Zephyr SMP server, sends an image for upgrade, and resets the emulated board. Afterwards MCUBoot does the swap and then I check that the image has been changed.

https://github.com/user-attachments/assets/afa3da5a-2892-4617-a4e9-53f95745ac48

The main advantage of this setup is that I can emulate the firmware on pipeline and require no hardware (nor BlueZ) as I use bumble as the HCI Controller stack (python). As I am using Bleak as my BLE frontend, the code is reusable and works with real and virtual hardware (renode). I will attempt to make it work with Android Emulator as I would like to do functional tests of a an Android App as SMP Client.

To make this work I developed a bumble [backend](https://github.com/hbldh/bleak/pull/1681) for bleak and added support for the Flash controller of the NRF52840 in renode.  In theory it should be doable to test this with other Zephyr boards as long as they have a UART (perhaps USB ?) that can be used as HCI Transport. Then in renode I expose the emulated HCI UART over a socket and connect it to Bumble (via Bleak).

**Comment by JPHutchins** [comment]
Absolutely incredible to be able to end-to-end test a DFU in an emulated environment! Including the bootloader and swap!

**Comment by vChavezB** [comment]
Just some updates on my side.

- I have made a repo with support files to enable the SoC Flash of the NRF52840  https://github.com/vChavezB/renode-nrf52840_flash.
- If I have a little more time (maybe after Holidays 🎅 ?) I will provide a demo with github action.

<img src="https://github.com/user-attachments/assets/62ed9564-f826-475d-adfb-84c52740d528" width="200">
